### ISSUE TITLE
FEAT: add fairness-aware Principal Component Analysis (preprocessing)

### ISSUE THREAD

**Issue by Thomasq99** [fairlearn/fairlearn#956]
PCA, one of the most common dimensionality reduction techniques might produce data representations that are inherently prone to unfairness. For example, an embedding of the data created by PCA might represent a certain race group much better than another.
 
In recent years a couple of alternatives have been proposed to ensure that disparity between sensitive groups is mitigated.
My inspiration/literature research is based on the following three papers:

[Fair PCA](https://arxiv.org/abs/1811.00103), [Efficient Fair PCA](https://arxiv.org/abs/1911.04931), [Multi Objective approach](https://arxiv.org/abs/2006.06137).

The first paper: Fair PCA aims to minimize the maximum additional reconstruction error for each population above the optimal n to d projection for that population alone. (the optimal projection is obtained by running PCA only on the sensitive group you are considering). See the image below for more details.

![image](https://user-images.githubusercontent.com/38749496/133259524-218023b0-1d2f-4815-ab62-8a6356a77d06.png)
 
The algorithm used to solve this problem relaxes the problem to a semidefinite optimization problem and solves the SDP, next they solve an LP designed to reduce the rank of said problem. This results in a relatively slow algorithm with as output a projection that has optimal fairness using at most d + k -1 features,, where d is the desired number of output features and k is the number of sensitive groups. (this algorithm is later changed to run slightly faster and require fewer additional features for the projection, see [updated algorithm](https://www.researchgate.net/publication/331429524_Fair_Dimensionality_Reduction_and_Iterative_Rounding_for_SDPs)).

The second paper [Efficient Fair PCA](https://arxiv.org/abs/1911.04931) uses a different loss function, namely that of pairwise disparity error. It is much alike the previous loss function however computes the difference in additional reconstruction error between each group. All of these errors are added as additional minimization objectives and the optimization then becomes to minimize the general reconstruction error as well as all additional objectives. They use an adaptive optimization technique based on gradient descent to solve this. It is guaranteed to converge to a Pareto stationary point and runs much faster than the previous discussed algorithm. Furthermore, it does not require any additional features and seems to work better for mitigating fairness by the experiments in their paper.

Finally, [Multi Objective approach](https://arxiv.org/abs/2006.06137) takes a different approach. Instead of creating a different projection matrix they aim to make the "fairest" selection of principal components. They measure fairness by taking the difference of reconstruction errors between both sensitive groups. They try to optimize the fairness measure and the total reconstruction error simultaneously by an evolutionary algorithm called [SPEA2](https://www.semanticscholar.org/paper/SPEA2%3A-Improving-the-strength-pareto-evolutionary-Zitzler-Laumanns/b13724cb54ae4171916f3f969d304b9e9752a57f). 

This method is nice for already deployed algorithms since it does not require to run a different PCA version. It only requires to reorder the principal components. 

I still have my doubts on the fairness measures accompanied by the algorithms and would like some discussion about that.
However, I do believe that implementing both  [Efficient Fair PCA](https://arxiv.org/abs/1911.04931), [Multi Objective approach](https://arxiv.org/abs/2006.06137) would be useful. Note that both focus on solving a multi objective optimization approach so I would need to check whether one of the two optimization algorithms is faster and whether they can be applied to both problems.

I would take this upon me so if someone can assign me to this, that would be great!

Let me know what you think and feel free to ask any questions!

Tasks:

- [ ] Think of a reasonable fairness measure
- [ ] Check whether optimization methods are applicable to both and check which one is faster
- [ ] implement efficient Fair PCA
- [ ] implement post processing fair PCA

**Comment by romanlutz** [comment]
This is awesome, thanks for the summary!

First of all, I would create separate issues for the various algorithms once you've decided which to implement. Starting from the metrics sounds like a very reasonable approach to me. For example, you could already evaluate standard (fairness-unaware) PCA that way!

#### Dependencies
If there are already implementations that exist for these algorithms I'd suggest checking their dependencies. We wouldn't want to introduce something heavyweight to Fairlearn. However, it might be fine as a soft dependency. I'm particularly thinking of packages to solve these optimization problems. Just something to keep in mind while you're working on this.

#### Metrics
In terms of measures I can think of the following (definitely not exhaustive):
- overall reconstruction error
- overall reconstruction loss of fair PCA compared to fairness-unaware PCA
- sensitive feature group-specific reconstruction error
- sensitive feature group-specific reconstruction loss of fair PCA compared to fairness-unaware PCA (for this latter one we'd have to decide whether to compare with PCA on just that group or PCA on the whole dataset, or alternatively do both)
- The most important metrics IMO are outcome based. It still surprises me that these papers usually don't run e2e analyses where they evaluate how using the method (or not) affects the outcomes per group as measured by typical performance metrics (grouped and ungrouped) as well as other potentially useful metrics (e.g., runtime). Perhaps that has something to do with paper length restrictions. In any case, these are all already available! We'd only have to set up an example to evaluate this end-to-end.
- any others mentioned in these papers?

#### Naming
I actually dislike the term "Fair PCA" a bit since it sort of implies that it is (or the results are) fair. As we've described in many places fairness is a sociotechnical concept, so the outcomes may be anything but fair. I can see why the considerations for a paper are somewhat different than for a project like Fairlearn, so perhaps it's time to rethink this before we add it to the package.
Some ideas:
- `FairnessAwarePCA`
- `GroupFairnessAwarePCA`
- `ReconstructionLossMinimizingPCA` (that's a mouthful...)
- something specific to the approach
- using author names (I personally dislike this since we'd be doing this for some techniques and not others; plus, we already cite the papers in the API reference and user guide, of course)

**Comment by Thomasq99** [comment]
Apologies for the amount of text below. I want to give as much information about the fairness measures to allow for a better discussion about them. 

All three papers below try to solve multi objective optimization problems. They try to minimize the overall reconstruction error while simultaneously maximizing a fairness metric.

For the fairness measures of the optimization problem, I was thinking about the ones proposed in the papers. They are the following:

The goal of the first two papers is to have the additional error compared to the optimal group-specific solution be equal for each group. For example, the additional error your dimensionality reduction algorithm imposes on females compared to the optimal solution for females should be equal to the additional error for males. Note that since the additional error compared to the optimal solution is used, it does not necessarily mean that females and males are equally well represented in the embedded space. The optimal solution for females could be much worse than for males.

**[Fair PCA](https://arxiv.org/abs/1811.00103)**:
Image below describes the fairness measure very well.
![image](https://user-images.githubusercontent.com/38749496/133588432-a32b9350-a0e4-47f7-97f9-9406879da756.png)

**[Efficient Fair PCA](https://arxiv.org/abs/1911.04931)**:
This method uses almost the same fairness measure, however, since optimal solutions do not always assign the same loss to each group, which is especially a problem when sensitive features have multiple groups, they add additional objectives to the multi objective loss function. In particular, they add the difference between the previously mentioned additional group-specific errors for each combination of groups (called pairwise disparity error). This ensures that the solution tries to minimize all pairwise disparity errors at the same time and addresses the aforementioned issue.

 **[Multi Objective approach](https://arxiv.org/abs/2006.06137)**:
This method simply tries to minimize the difference in reconstruction error of each group compared to each other group. This thus does not take into account that the optimal solution for one group might be very different than another group.

Currently, I am debating which of the three makes the most sense. 

Compared to the first fairness measure (paper 1) the second fairness measure (paper 2) is, in my opinion, better than the first one since it is designed to alleviate the issue of the first paper. 

The fairness measure of the third paper does not take into account that the optimal solution for each group might have a vastly different reconstruction error. This can be "problematic". Let's say the optimal reconstruction error for females is ten times as high as males. If your desired solution measures fairness in the way paper 3 does it might not improve the reconstruction error for males since it only cares about the difference between the two. Improving the reconstruction error for males even though the female reconstruction error is higher, yet optimal for that group, would decrease the fairness. This results in an overall reconstruction error which is worse. 

The "problem" with the fairness metric in paper 3 is thus that the way fairness is measured might cause the algorithm to not improve on overall reconstruction error because the fairness measure gets worse, even though the optimal reconstruction error for a certain group is already met. The fairness metric in paper 2 does not have this "problem" since it takes into account the additional reconstruction error compared to the optimal group-specific reconstruction error. I am, however, doubting how fair a solution is which has a very low reconstruction error for males compared to females even though the optimal reconstruction error for females is much lower than males. Paper 2 would call it a fair solution since the additional errors might be very low, however, the reconstruction errors between groups still differ greatly. I doubt whether this would result in a fair decomposition. It seems to me that in this case, it is impossible to do PCA in a fair way since males will be much better represented than females regardless.

What are your thoughts on the advantages/disadvantages between these fairness measures? Do you agree with the problems sketched in the above paragraph or do you think I am missing something?

Maybe a nice alternative would be to implement both fairness measures and allow the user to choose. Of course, I would evaluate both of them and add my findings in the documentation. Do you agree this is the way to go?

**Comment by MiroDudik** [comment]
@Thomasq99: Can you spell out more precisely different objectives / fairness measures? It would be good if you presented them in the form that makes it clear how to evaluate them with more than two groups (unless the generalization to multiple groups is obvious). Feel free to just copy-paste the relevant snippets from the papers.

As things stand, fairlearn supports both worst-case performance metrics called `*_group_min()` and `*_group_max()` for the smallest/largest metric value across slices of data corresponding to different group, as well as the disparity metrics called `*_difference` for the largest (pairwise) difference and `*_ratio` for the smallest (pairwise) ratio (see [[here]](https://fairlearn.org/v0.7.0/user_guide/assessment.html#scalar-results-from-metricframe)).

In the absence of more information (what are the downstream harms caused by PCA?), the goals of generic preprocessing seem to fall into the [quality-of-service setting](https://fairlearn.org/v0.7.0/user_guide/fairness_in_machine_learning.html#types-of-harms), for which worst-case performance fairness metrics are more appropriate, so my hunch would be to start with the "Fair PCA" algorithm, but like @romanlutz said, we may want to rename it to be more descriptive of the kind of fairness it seeks to enforce. The idea is quite close to what we call [bounded group loss](https://fairlearn.org/v0.7.0/user_guide/mitigation.html#bounded-group-loss), but in "Fair PCA" it is presented as an objective, whereas in our reductions it appears as a constraint. We could possibly call it "MinMaxPCA" since it minimizes the worst-case reconstruction error among low-rank matrices?

By the way, we already have a preprocessing technique that is being applied to matrices and uses reconstruction error as the underlying metric, so you may want to use it for an inspiration for what the API might look like: [CorrelationRemover](https://fairlearn.org/v0.7.0/api_reference/fairlearn.preprocessing.html).

**Comment by romanlutz** [comment]
FWIW I am always in favor of providing as many metrics as possible, explaining their benefits and drawbacks (e.g., through a user guide or example), and then letting users choose which ones they care about. IMO the downstream harms are the most important, so we should really be looking at the fairness metrics for the whole system (not just the preprocessing step). Having a few specific ones for preprocessing doesn't hurt, of course, but we shouldn't get into the habit of thinking that it's sufficient to minimize the max reconstruction error of all groups (to mention one particular metric).

I'd definitely consider creating a very small example (you can check the existing ones in case one of those suits you) where you apply "normal" PCA, `CorrelationRemover`, and then subsequently the other techniques as you add them. Then you can measure these new metrics as well as the downstream effects on fairness metrics.

@hildeweerts do feel free to chime in! 🙂

**Comment by hildeweerts** [comment]
#### Fairness metrics for evaluating PCA
From the discussion above, it seems to me that a fairness metric tailored towards PCA could consider the reconstruction error across groups (but please correct me if I'm wrong). However, I fully agree with @romanlutz and @MiroDudik that downstream model performance should be leading in assessing fairness (and should be included in the user guide section for fairness-aware PCA). This also makes me think that the problems described with the "optimal" reconstruction error may not be *super* relevant, as it's the final model performance that we care about the most - but let me know if anybody disagrees!

#### Fairness-aware PCA algorithms
I like @MiroDudik 's suggestion to summarize and specific the objectives more precisely for each of the suggested fairness-aware PCA algorithms. I think that would make it easier to compare them.

**Comment by Thomasq99** [comment]
Update:

Currently I am implementing the methods and loss functions to evaluate them on downstream performance. I will get back to you soon!

**Comment by romanlutz** [comment]
FWIW I didn't mean to make you run entire studies against benchmarks. I merely meant that the most meaningful metrics are those that capture downstream harms so we don't need to spend a ton of time implementing more metrics in addition to reconstruction error that only capture the results of preprocessing (since these don't necessarily imply harm). If you want to run extensive experiments I'm all for it, but I'd vouch for the inclusion of popular methods even based on face validity of the method (as @MiroDudik and others mentioned yesterday in the community call). Anyway, let us know if you need help in some way!

**Comment by Thomasq99** [comment]
Don't worry! I am interested in doing this study myself. It seemed like a nice internship topic to me!
